scipy/sparse/linalg/eigen/arpack/arpack.py

```diff
@@ -1790,40 +1800,60 @@ def svds(A, k=6, ncv=None, tol=0, which='LM', v0=None,
             X_dot = A.matvec
             X_matmat = A.matmat
             XH_dot = A.rmatvec
+            XH_mat = A.rmatmat
         else:
             X_dot = A.rmatvec
+            X_matmat = A.rmatmat
             XH_dot = A.matvec
+            XH_mat = A.matmat
 
             dtype = getattr(A, 'dtype', None)
             if dtype is None:
-                dtype = A.dot(np.zeros([m,1])).dtype
-
-            # A^H * V; works around lack of LinearOperator.adjoint.
-            # XXX This can be slow!
-            def X_matmat(V):
-                out = np.empty((V.shape[1], m), dtype=dtype)
-                for i, col in enumerate(V.T):
-                    out[i, :] = A.rmatvec(col.reshape(-1, 1)).T
-                return out.T
+                dtype = A.dot(np.zeros([m, 1])).dtype
 
     else:
         if n > m:
             X_dot = X_matmat = A.dot
-            XH_dot = _herm(A).dot
+            XH_dot = XH_mat = _herm(A).dot
         else:
-            XH_dot = A.dot
+            XH_dot = XH_mat = A.dot
             X_dot = X_matmat = _herm(A).dot
 
     def matvec_XH_X(x):
         return XH_dot(X_dot(x))
 
+    def matmat_XH_X(x):
+        return XH_mat(X_matmat(x))
+
     XH_X = LinearOperator(matvec=matvec_XH_X, dtype=A.dtype,
+                          matmat=matmat_XH_X,
                           shape=(min(A.shape), min(A.shape)))
 
     # Get a low rank approximation of the implicitly defined gramian matrix.
     # This is not a stable way to approach the problem.
-    eigvals, eigvec = eigsh(XH_X, k=k, tol=tol ** 2, maxiter=maxiter,
-                                  ncv=ncv, which=which, v0=v0)
+    if solver == 'lobpcg':
+
+        if which == 'LM':
+            largest = True
+        elif which == 'SM':
+            largest = False
+        else:
+            raise ValueError("which must be either 'LM' or 'SM'.")
+
+        if k == 1 and v0 is not None:
+            X = np.reshape(v0, (-1, 1))
+        else:
+            X = np.random.RandomState(52).randn(min(A.shape), k)
+
+        eigvals, eigvec = lobpcg(XH_X, X, tol=tol ** 2, maxiter=maxiter,
```

Comment:
Does lobpcg behave similarly as arpack with the default value `tol=0`? (ARPACK does `if tol == 0: tol = macheps`)

Reply:
Yes, the current lobpcg revision in master does 
```
residualTolerance = tol
...
if (residualTolerance is None) or (residualTolerance <= 0.0):
    residualTolerance = np.sqrt(1e-15) * n
```
so tol == 0 is treated sensibly